File: styleable-views/src/main/java/com/github/cthawanapong/view/StyleableButton.java
package com.github.cthawanapong.view;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.github.cthawanapong.model.BundleSavedState;
import com.github.cthawanapong.styleableviews.R;

/**
 * Created by Chavit Thawanapong on 4/25/2017 AD.
 * Email: c.thawanapong@gmail.com
 */

public class StyleableButton extends FrameLayout implements View.OnClickListener {
    private static final String TAG = StyleableButton.class.getSimpleName();

    private static final String ARG_CHILD_STATES = "childrenStates";
    private static final String ARG_BUTTON_TEXT = "ARG_BUTTON_TEXT";
    private static final String ARG_BUTTON_PADDING = "ARG_BUTTON_PADDING";
    private static final String ARG_TEXT_SIZE = "ARG_TEXT_SIZE";
    private static final String ARG_BUTTON_BACKGROUND = "ARG_BUTTON_BACKGROUND";
    private static final String ARG_TEXT_COLOR = "ARG_TEXT_COLOR";
    private static final String ARG_DRAWABLE_PADDING = "ARG_DRAWABLE_PADDING";
    private static final String ARG_HORIZONTAL_PADDING = "ARG_HORIZONTAL_PADDING";
    private static final String ARG_BORDER_COLOR = "ARG_BORDER_COLOR";
    private static final String ARG_CORNER_RADIUS = "ARG_CORNER_RADIUS";
    private static final String ARG_ELEVATION = "ARG_ELEVATION";

    //View Members
    private StyleableNoPaddingCardView mLayoutCard;
    private ViewGroup mLayoutButton;
    private ViewGroup mLayoutText;
    private ViewGroup mLayoutWrapper;
    private StyleableTextView mTextView;

    //Data Members
    private String mButtonText;
    private int mPadding;
    private float mTextSize;
    private int mBackgroundColor;
    private int mTextColor;
    private Drawable mLeftIcon;
    private int mButtonDrawablePadding;
    private int mButtonHorizontalPadding;
    private int mBorderColor;
    private int mCornerRadius;
    private int mElevation;
    private OnClickListener mListener;

    public StyleableButton(Context context) {
        super(context);
        initInflate();
        initInstance(context, null);
    }

    public StyleableButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        initInflate();
        initInstance(context, attrs);
    }

    public StyleableButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initInflate();
        initInstance(context, attrs);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public StyleableButton(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initInflate();
        initInstance(context, attrs);
    }

    private void initInflate() {
        //Inflate Layout
        inflate(getContext(), R.layout.view_button, this);
    }

    private void initInstance(Context context, AttributeSet attrs) {
        mLayoutCard = (StyleableNoPaddingCardView) findViewById(R.id.layout_border);
        mLayoutButton = (ViewGroup) findViewById(R.id.layout_button);
        mLayoutText = (ViewGroup) findViewById(R.id.layout_text);
        mLayoutWrapper = (ViewGroup) findViewById(R.id.layout_wrapper);
        mTextView = (StyleableTextView) findViewById(R.id.txt_view);
        mButtonText = "";

        if (attrs != null) {
            TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.StyleableButton, 0, 0);
            try {
                mButtonText = ta.getString(R.styleable.StyleableButton_buttonText);
                mPadding = ta.getDimensionPixelSize(R.styleable.StyleableButton_buttonPadding, 0);
                mTextSize = ta.getDimension(R.styleable.StyleableButton_buttonTextSize, getContext().getResources().getDimension(R.dimen.default_button_text_size));
                mBackgroundColor = ta.getColor(R.styleable.StyleableButton_buttonBackgroundColor, Color.WHITE);
                mTextColor = ta.getColor(R.styleable.StyleableButton_buttonTextColor, ContextCompat.getColor(getContext(), R.color.default_text_color));
                mLeftIcon = ta.getDrawable(R.styleable.StyleableButton_buttonLeftDrawable);
                mButtonDrawablePadding = ta.getDimensionPixelSize(R.styleable.StyleableButton_buttonDrawablePadding, 0);
                mButtonHorizontalPadding = ta.getDimensionPixelSize(R.styleable.StyleableButton_buttonHorizontalPadding, 0);
                mBorderColor = ta.getColor(R.styleable.StyleableButton_buttonBorderColor, Color.BLACK);
                mCornerRadius = ta.getDimensionPixelSize(R.styleable.StyleableButton_buttonCornerRadius, 0);
                mElevation = ta.getDimensionPixelSize(R.styleable.StyleableButton_buttonElevation, 0);
            } finally {
                ta.recycle();
            }
        }

        mLayoutButton.setPadding(mPadding, mPadding, mPadding, mPadding);
        mLayoutText.setBackgroundColor(mBackgroundColor);
        mTextView.setTextColor(mTextColor);
        mTextView.setTextSize(TypedValue.COMPLEX_UNIT_PX, mTextSize);
        mTextView.setCompoundDrawablesWithIntrinsicBounds(mLeftIcon, null, null, null);
        mTextView.setCompoundDrawablePadding(mButtonDrawablePadding);
        mLayoutCard.setCardBackgroundColor(mBorderColor);
        mLayoutCard.setRadius(mCornerRadius);
        mLayoutCard.setCardElevation(mElevation);
        mLayoutWrapper.setPadding(mButtonHorizontalPadding, 0, mButtonHorizontalPadding, 0);
        mLayoutWrapper.setOnClickListener(this);
        setDisplayButtonText();
    }

    private void setDisplayButtonText() {
        mTextView.setText(mButtonText);
    }

    @Override
    protected void dispatchSaveInstanceState(SparseArray<Parcelable> container) {
        dispatchFreezeSelfOnly(container);
    }

    @Override
    protected void dispatchRestoreInstanceState(SparseArray<Parcelable> container) {
        dispatchThawSelfOnly(container);
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        Parcelable superState = super.onSaveInstanceState();

        //Save Child State
        int id;
        Bundle childrenStates = new Bundle();
        for (int i = 0; i < getChildCount(); i++) {
            id = getChildAt(i).getId();
            if (id != 0) {
                SparseArray childrenState = new SparseArray();
                getChildAt(i).saveHierarchyState(childrenState);
                childrenStates.putSparseParcelableArray(String.valueOf(id), childrenState);
            }
        }
        Bundle bundle = new Bundle();
        bundle.putBundle(ARG_CHILD_STATES, childrenStates);
        bundle.putString(ARG_BUTTON_TEXT, mButtonText);
        bundle.putInt(ARG_BUTTON_PADDING, mPadding);
        bundle.putFloat(ARG_TEXT_SIZE, mTextSize);
        bundle.putInt(ARG_BUTTON_BACKGROUND, mBackgroundColor);
        bundle.putInt(ARG_TEXT_COLOR, mTextColor);
        bundle.putInt(ARG_DRAWABLE_PADDING, mButtonDrawablePadding);
        bundle.putInt(ARG_HORIZONTAL_PADDING, mButtonHorizontalPadding);
        bundle.putInt(ARG_BORDER_COLOR, mBorderColor);
        bundle.putInt(ARG_CORNER_RADIUS, mCornerRadius);
        bundle.putInt(ARG_ELEVATION, mElevation);

        //Save it to Parcelable
        BundleSavedState ss = new BundleSavedState(superState);
        ss.setBundle(bundle);
        return ss;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        BundleSavedState ss = (BundleSavedState) state;
        super.onRestoreInstanceState(ss.getSuperState());

        //Restore SparseArray

        Bundle bundle = ss.getBundle();
        Bundle childStates = bundle.getBundle(ARG_CHILD_STATES);

        //Restore Children's state
        int id;
        for (int i = 0; i < getChildCount(); i++) {
            id = getChildAt(i).getId();
            if (id != 0) {
                if (childStates.containsKey(String.valueOf(id))) {
                    SparseArray childrenState =
                            childStates.getSparseParcelableArray(String.valueOf(id));
                    getChildAt(i).restoreHierarchyState(childrenState);
                }
            }
        }

        //Restore State Here
        this.mButtonText = bundle.getString(ARG_BUTTON_TEXT);
        this.mPadding = bundle.getInt(ARG_BUTTON_PADDING);
        this.mTextSize = bundle.getFloat(ARG_TEXT_SIZE);
        this.mBackgroundColor = bundle.getInt(ARG_BUTTON_BACKGROUND);
        this.mTextColor = bundle.getInt(ARG_TEXT_COLOR);
        this.mButtonDrawablePadding = bundle.getInt(ARG_DRAWABLE_PADDING);
        this.mButtonHorizontalPadding = bundle.getInt(ARG_HORIZONTAL_PADDING);
        this.mBorderColor = bundle.getInt(ARG_BORDER_COLOR);
        this.mCornerRadius = bundle.getInt(ARG_CORNER_RADIUS);
        this.mElevation = bundle.getInt(ARG_ELEVATION);
    }

    public void setListener(OnClickListener mListener) {
        this.mListener = mListener;
    }

    public void setButtonText(String mButtonText) {
        this.mButtonText = mButtonText;

        setDisplayButtonText();
    }

    public String getButtonText() {
        return mButtonText;
    }

    public void setBackgroundColor(int color) {
        mBackgroundColor = color;
        mLayoutText.setBackgroundColor(mBackgroundColor);
    }

    public void setBorderColor(int color) {
        mBorderColor = color;
        mLayoutCard.setCardBackgroundColor(mBorderColor);
    }

    @Override
    public void onClick(View v) {
        if (mListener != null && v instanceof ViewGroup) {
            mListener.onClick((ViewGroup) v);
        }
    }

    public interface OnClickListener {
        void onClick(ViewGroup viewGroup);
    }
}

File: styleable-views/src/main/java/com/github/cthawanapong/view/StyleableButton.kt
package com.github.cthawanapong.view

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.Parcelable
import android.support.v4.content.ContextCompat
import android.support.v7.content.res.AppCompatResources
import android.support.v7.widget.CardView
import android.util.AttributeSet
import android.util.SparseArray
import android.util.TypedValue
import android.view.View
import android.widget.Button
import com.github.cthawanapong.model.BundleSavedState
import com.github.cthawanapong.styleableviews.R
import kotlinx.android.synthetic.main.view_styleable_button.view.*


/**
 * Created by CThawanapong on 27/1/2018 AD.
 * Email: c.thawanapong@gmail.com
 */
class StyleableButton @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0, defStyleRes: Int = 0
) : CardView(context, attrs, defStyleAttr) {
    companion object {
        @JvmStatic
        private val TAG = StyleableButton::class.java.simpleName

        @JvmStatic
        private val ARG_CHILD_STATES = "childrenStates"
        @JvmStatic
        private val ARG_BUTTON_TEXT = "ARG_BUTTON_TEXT"
        @JvmStatic
        private val ARG_TEXT_SIZE = "ARG_TEXT_SIZE"
        @JvmStatic
        private val ARG_BUTTON_BACKGROUND = "ARG_BUTTON_BACKGROUND"
        @JvmStatic
        private val ARG_TEXT_COLOR = "ARG_TEXT_COLOR"
        @JvmStatic
        private val ARG_DRAWABLE_PADDING = "ARG_DRAWABLE_PADDING"
        @JvmStatic
        private val ARG_HORIZONTAL_PADDING = "ARG_HORIZONTAL_PADDING"
        @JvmStatic
        private val ARG_VERTICAL_PADDING = "ARG_VERTICAL_PADDING"
        @JvmStatic
        private val ARG_BORDER_COLOR = "ARG_BORDER_COLOR"
        @JvmStatic
        private val ARG_CORNER_RADIUS = "ARG_CORNER_RADIUS"
        @JvmStatic
        private val ARG_ELEVATION = "ARG_ELEVATION"
    }

    //Data Members
    var buttonText: String = ""
        set(value) {
            field = value
            setInternalButtonText(field)
        }
    var buttonTextSize: Float = 0.toFloat()
        set(value) {
            field = value
            setInternalTextSize(field)
        }
    var buttonBackgroundColor: Int = 0
        set(value) {
            field = value
            setInternalButtonBackgroundColor(field)
        }
    var buttonTextColor: Int = 0
        set(value) {
            field = value
            setInternalButtonTextColor(field)
        }
    var buttonLeftIcon: Drawable? = null
        set(value) {
            field = value
            setInternalButtonLeftIcon(field)
        }
    var buttonDrawablePadding: Int = 0
        set(value) {
            field = value
            setInternalButtonDrawablePadding(field)
        }
    var buttonHorizontalPadding: Int = 0
        set(value) {
            field = value
            setInternalButtonPadding()
        }
    var buttonVerticalPadding: Int = 0
        set(value) {
            field = value
            setInternalButtonPadding()
        }
    var buttonBorderColor: Int = 0
        set(value) {
            field = value
            setInternalButtonBorderColor(field)
        }
    var buttonCornerRadius: Int = 0
        set(value) {
            field = value
            setInternalButtonCornerRadius(field)
        }
    var buttonElevation: Int = 0
        set(value) {
            field = value
            setInternalButtonElevation(field)
        }
    val styleableButton: Button
        get() = button

    init {
        initInflate()
        initInstance(context, attrs, defStyleAttr)
    }

    private fun initInflate() {
        View.inflate(context, R.layout.view_styleable_button, this)
    }

    private fun initInstance(context: Context, attrs: AttributeSet?, defStyleAttr: Int) {
        attrs?.let {
            val ta = context.obtainStyledAttributes(it, R.styleable.StyleableButton)

            with(ta) {
                buttonText = ta.getString(R.styleable.StyleableButton_styleableButtonText)
                buttonTextSize = ta.getDimensionPixelSize(R.styleable.StyleableButton_styleableButtonTextSize, context.resources.getDimensionPixelSize(R.dimen.styleable_default_button_text_size)).toFloat()
                buttonBackgroundColor = ta.getColor(R.styleable.StyleableButton_styleableButtonBackgroundColor, Color.TRANSPARENT)
                buttonTextColor = ta.getColor(R.styleable.StyleableButton_styleableButtonTextColor, ContextCompat.getColor(context, R.color.default_text_color))
                val buttonLeftIconRes = ta.getResourceId(R.styleable.StyleableButton_styleableButtonLeftDrawable, 0)
                if (buttonLeftIconRes != 0) {
                    buttonLeftIcon = AppCompatResources.getDrawable(context, buttonLeftIconRes)
                }
                buttonDrawablePadding = ta.getDimensionPixelSize(R.styleable.StyleableButton_styleableButtonDrawablePadding, 0)
                buttonHorizontalPadding = ta.getDimensionPixelSize(R.styleable.StyleableButton_styleableButtonHorizontalPadding, 0)
                buttonVerticalPadding = ta.getDimensionPixelSize(R.styleable.StyleableButton_styleableButtonVerticalPadding, context.resources.getDimension(R.dimen.styleable_default_button_horizontal_padding).toInt())
                buttonBorderColor = ta.getColor(R.styleable.StyleableButton_styleableButtonBorderColor, Color.TRANSPARENT)
                buttonCornerRadius = ta.getDimensionPixelSize(R.styleable.StyleableButton_styleableButtonCornerRadius, 0)
                buttonElevation = ta.getDimensionPixelSize(R.styleable.StyleableButton_styleableButtonElevation, 0)
            }

            ta.recycle()
        }
    }

    private fun setInternalButtonText(buttonText: String) {
        button.text = buttonText
    }

    private fun setInternalTextSize(textSize: Float) {
        button.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize)
    }

    private fun setInternalButtonBackgroundColor(backgroundColor: Int) {
        layoutText.setCardBackgroundColor(backgroundColor)
    }

    private fun setInternalButtonTextColor(color: Int) {
        button.setTextColor(color)
    }

    private fun setInternalButtonLeftIcon(leftIcon: Drawable?) {
        button.setCompoundDrawablesWithIntrinsicBounds(leftIcon, null, null, null)
    }

    private fun setInternalButtonDrawablePadding(padding: Int) {
        button.compoundDrawablePadding = padding
    }

    private fun setInternalButtonPadding() {
        button.setPadding(buttonHorizontalPadding, buttonVerticalPadding, buttonHorizontalPadding, buttonVerticalPadding)
    }

    private fun setInternalButtonBorderColor(color: Int) {
        setCardBackgroundColor(color)
    }

    private fun setInternalButtonCornerRadius(newRadius: Int) {
        layoutText.radius = newRadius.toFloat()
        radius = newRadius.toFloat()
    }

    private fun setInternalButtonElevation(elevation: Int) {
        cardElevation = elevation.toFloat()
    }

    override fun dispatchSaveInstanceState(container: SparseArray<Parcelable>) {
        dispatchFreezeSelfOnly(container)
    }

    override fun dispatchRestoreInstanceState(container: SparseArray<Parcelable>) {
        dispatchThawSelfOnly(container)
    }

    override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState()

        //Save Child State
        var id: Int
        val childrenStates = Bundle()
        for (i in 0 until childCount) {
            id = getChildAt(i).id
            if (id != 0) {
                val childrenState = SparseArray<Parcelable>()
                getChildAt(i).saveHierarchyState(childrenState)
                childrenStates.putSparseParcelableArray(id.toString(), childrenState)
            }
        }

        val saveBundle = Bundle().apply {
            putBundle(ARG_CHILD_STATES, childrenStates)
            putString(ARG_BUTTON_TEXT, buttonText)
            putFloat(ARG_TEXT_SIZE, buttonTextSize)
            putInt(ARG_BUTTON_BACKGROUND, buttonBackgroundColor)
            putInt(ARG_TEXT_COLOR, buttonTextColor)
            putInt(ARG_DRAWABLE_PADDING, buttonDrawablePadding)
            putInt(ARG_HORIZONTAL_PADDING, buttonHorizontalPadding)
            putInt(ARG_VERTICAL_PADDING, buttonVerticalPadding)
            putInt(ARG_BORDER_COLOR, buttonBorderColor)
            putInt(ARG_CORNER_RADIUS, buttonCornerRadius)
            putInt(ARG_ELEVATION, buttonElevation)
        }

        //Save it to Parcelable
        return BundleSavedState(superState)
                .apply {
                    bundle = saveBundle
                }
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        val ss = state as BundleSavedState
        super.onRestoreInstanceState(ss.superState)

        //Restore SparseArray

        val bundle = ss.bundle
        val childStates = bundle.getBundle(ARG_CHILD_STATES)

        //Restore Children's state
        var id: Int
        for (i in 0 until childCount) {
            id = getChildAt(i).id
            if (id != 0) {
                if (childStates.containsKey(id.toString())) {
                    val childrenState = childStates.getSparseParcelableArray<Parcelable>(id.toString())
                    getChildAt(i).restoreHierarchyState(childrenState)
                }
            }
        }

        //Restore State Here
        buttonText = bundle.getString(ARG_BUTTON_TEXT)
        buttonTextSize = bundle.getFloat(ARG_TEXT_SIZE)
        buttonBackgroundColor = bundle.getInt(ARG_BUTTON_BACKGROUND)
        buttonTextColor = bundle.getInt(ARG_TEXT_COLOR)
        buttonDrawablePadding = bundle.getInt(ARG_DRAWABLE_PADDING)
        buttonHorizontalPadding = bundle.getInt(ARG_HORIZONTAL_PADDING)
        buttonVerticalPadding = bundle.getInt(ARG_VERTICAL_PADDING)
        buttonBorderColor = bundle.getInt(ARG_BORDER_COLOR)
        buttonCornerRadius = bundle.getInt(ARG_CORNER_RADIUS)
        buttonElevation = bundle.getInt(ARG_ELEVATION)
    }
}